Renamed DictionaryLiteral to KeyValuePairs

masters3d opened the pull request:
Implementation for https://github.com/apple/swift-evolution/pull/850

"There is no strong motivation to deprecate. The type does not produce
active harm. Instead, it adds measurable (if small) utility and will be
part of the ABI. A sensible renaming mitigates the most problematic
issue with the type."

masters3d commented:
@swift-ci Please smoke test

nh7a commented:
The proposal is suggesting `KeyValueList`, not `KeyValueArray`, though.

benrimmington commented on include/swift/AST/Expr.h:
```diff
@@ -149,7 +149,7 @@ class alignas(8) Expr {
     /// Number of comma source locations.
     NumCommas : 32 - 1 - NumExprBits,
-    /// Number of entries in the collection. If this is a DictionaryLiteral,
+    /// Number of entries in the collection. If this is a KeyValueArray,
```
I think this refers to `DictionaryExpr` (which is a subclass of `CollectionExpr`).

benrimmington replied:
The tuples are probably those given to any [`init(dictionaryLiteral:)`](https://github.com/apple/swift/blob/b6e637e03d012121c59c02aab7915f563401814e/stdlib/public/core/CompilerProtocols.swift#L699-L706) initializer. I don't think `AST/Expr.h` would need to have any knowledge of the `KeyValueArray` type.

masters3d replied:
This is part of the confusion and reason behind the rename. I believe while `AST/Expr.h` has not much to do with concrete types, the comment seems to be referring to the concrete type. Since  this comment is not user facing I am not too worried about it.  You are welcome to change it back. I added you as a collaborator to my branch. Are you able to start the CI? Thanks!

benrimmington replied:
I haven't got commit access, so I'm not able to trigger CI testing.

But the [swift-package-manager](https://github.com/apple/swift-package-manager/search?q=DictionaryLiteral&type=Code) didn't compile last time, did it?

You'll either need to:

- use [cross repository testing](https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#cross-repository-testing) with another PR, or
- wait until the compiler supports typealias extensions.

I also wondered if a [generic typealias](https://github.com/apple/swift-evolution/blob/master/proposals/0048-generic-typealias.md) might help, but it didn't seem to (with a recent Swift 4.2 snapshot).

benrimmington commented:
`DictionaryLiteral` is used in:

* [benchmark/single-source/RomanNumbers.swift](https://github.com/apple/swift/blob/master/benchmark/single-source/RomanNumbers.swift)
* [benchmark/single-source/StringMatch.swift](https://github.com/apple/swift/blob/master/benchmark/single-source/StringMatch.swift)
* [stdlib/private/StdlibUnittest/TypeIndexed.swift](https://github.com/apple/swift/blob/master/stdlib/private/StdlibUnittest/TypeIndexed.swift)
* [stdlib/public/core/CompilerProtocols.swift](https://github.com/apple/swift/blob/b6e637e03d012121c59c02aab7915f563401814e/stdlib/public/core/CompilerProtocols.swift#L665-L668)
* [test/stdlib/DictionaryLiteral.swift](https://github.com/apple/swift/blob/master/test/stdlib/DictionaryLiteral.swift)

masters3d commented:
@swift-ci Please test source compatibility

aciidb0mb3r commented:
@swift-ci Please test source compatibility

aciidb0mb3r commented:
@swift-ci please test source compatibility

masters3d commented:
none of the failures I saw are related https://ci.swift.org/job/swift-PR-source-compat-suite/1219/artifact/swift-source-compat-suite/

Aside from the Package Manager, most people do not seem to know or extend the type so I think they will not be any source compatibility. 
https://github.com/search?p=2&q="extension+DictionaryLiteral+where"&type=Code&utf8=✓

masters3d commented:
Updated. Not sure what to do about [stdlib-stable.json](https://github.com/apple/swift/blob/master/test/api-digester/stdlib-stable.json

lattner commented:
Great works @masters3d.  The patch LGTM, but it would be good to get a standard library reviewer like @airspeedswift or @dabrahams to take a look.

airspeedswift reviewed (approved):
LGTM with the exception of the question of the deprecation.

airspeedswift commented:
This all looks good to me. However, I am not sure about the deprecation in 4.2. While we could do it in 4.2, I think that's a little aggressive.

My preference would be to add the new name as a typealias in 4.2 now, and to deprecate the old name in 5.0. That way, people will be able to update their code bases in advance, and once 5.0 comes around, we can deprecate the old name in a way that means users will be able to compile without warnings with both the 4.2 and 5.0 compiler.

lattner commented:
I don't think this type is widely used (does it appear in the source compat suite at all?) so I don't think that warning in 4.2 will be too aggressive.  That said, your approach is also totally fine with me.

benrimmington commented on CHANGELOG.md:
```diff
@@ -57,6 +57,12 @@ Swift 5.0
 Swift 4.2
 ---------
 
+* [SE-0214][]:
```
You'll need to add a reference to the end of this file. The missing references are:

```
[SE-0213]: <https://github.com/apple/swift-evolution/blob/master/proposals/0213-literal-init-via-coercion.md>
[SE-0214]: <https://github.com/apple/swift-evolution/blob/master/proposals/0214-DictionaryLiteral.md>
[SE-0215]: <https://github.com/apple/swift-evolution/blob/master/proposals/0215-conform-never-to-hashable-and-equatable.md>
[SE-0216]: <https://github.com/apple/swift-evolution/blob/master/proposals/0216-dynamic-callable.md>
```

masters3d replied:
The others are not mention on the document yet so I only added [SE-0213]

benrimmington replied:
This changelog entry will probably need to be moved to the Swift 5.0 section.

The other typealias from #17711 could be documented here (Swift 4.2) instead.

benrimmington commented:
There's a documentation comment to update in:

* [stdlib/public/core/CompilerProtocols.swift](https://github.com/apple/swift/blob/50fb3b8496f9501c95326a1f9e33638e56797fb0/stdlib/public/core/CompilerProtocols.swift#L665-L668)

The `DictionaryLiteral` typealias will be used in:

* [benchmark/single-source/RomanNumbers.swift](https://github.com/apple/swift/blob/50fb3b8496f9501c95326a1f9e33638e56797fb0/benchmark/single-source/RomanNumbers.swift#L28)
* [stdlib/private/StdlibUnittest/TypeIndexed.swift](https://github.com/apple/swift/blob/50fb3b8496f9501c95326a1f9e33638e56797fb0/stdlib/private/StdlibUnittest/TypeIndexed.swift#L101)

masters3d commented:
Updated.

masters3d commented:
Can this be merged?

airspeedswift commented:
@swift-ci please test and merge

airspeedswift commented:
Hi @masters3d can you rebase this PR and we can try again? Thanks!

masters3d committed:
renamed DictionaryLiteral to KeyValuePairs per SE-0214

masters3d committed:
renamed DictionaryLiteral type tests to KeyValuePairs

masters3d commented:
done.

airspeedswift commented:
@swift-ci please test

benrimmington commented on stdlib/public/core/Mirror.swift:
```diff
@@ -499,67 +499,70 @@ extension Mirror {
 
-/// `Collection` conformance that allows `DictionaryLiteral` to
+@available(swift, deprecated: 5.0, renamed: "KeyValuePairs")
+public typealias DictionaryLiteral = KeyValuePairs
```
Similar to #17711, should this be a generic typealias?

```diff
 @available(swift, deprecated: 5.0, renamed: "KeyValuePairs")
-public typealias DictionaryLiteral = KeyValuePairs
+public typealias DictionaryLiteral<Key, Value> = KeyValuePairs<Key, Value>
```

masters3d replied:
I don't have a preference either way. @benrimmington you have write access to the branch. Please feel free to update it.

benrimmington replied:
@masters3d I've rebased and tested the changes locally, but I can't push to your branch. Please try adding me as a collaborator again.

benrimmington replied:
@masters3d I'm probably doing something wrong, but authentication fails on the command line (`git push masters3d +renamed-to-KeyValueArray`).

I also tried pushing from Xcode, but it fails because your branch is 1105 commits behind `apple:master`. If you can rebase your branch from upstream, I'll try again.

masters3d replied:
https://help.github.com/articles/pushing-to-a-remote/![image](https://user-images.githubusercontent.com/6539412/43377191-40e3c51a-9374-11e8-8afd-9ca18412d848.png)

You should be able to push, try pushing a branch with a new name (that should work)

benrimmington replied:
@masters3d If I create a new branch in your repository, it won't update this pull request. I've pushed to [`benrimmington:renamed-to-KeyValueArray`](https://github.com/benrimmington/swift/tree/renamed-to-KeyValueArray) for now, but it will be simpler to ignore that and make your own changes instead.

benrimmington committed:
[SE-0214] Move changelog entry (Swift 4.2 => 5.0)

benrimmington committed:
[SE-0214] Update comment in AST/Expr.h

benrimmington committed:
[SE-0214] Use generic typealias

See also <https://github.com/apple/swift/pull/17711>

masters3d commented:
It looks like you need to accept the invitation to collaborate.
https://github.com/masters3d/swift/invitations

On Sun, Jul 29, 2018 at 6:30 AM, Ben Rimmington <notifications@github.com>
wrote:

> *@benrimmington* commented on this pull request.
> ------------------------------
>
> In stdlib/public/core/Mirror.swift
> <https://github.com/apple/swift/pull/16577#discussion_r205973877>:
>
> >    public init(dictionaryLiteral elements: (Key, Value)...) {
>      self._elements = elements
>    }
>    @usableFromInline // FIXME(sil-serialize-all)
>    internal let _elements: [(Key, Value)]
>  }
>
> -/// `Collection` conformance that allows `DictionaryLiteral` to
> +@available(swift, deprecated: 5.0, renamed: "KeyValuePairs")
> +public typealias DictionaryLiteral = KeyValuePairs
>
> @masters3d <https://github.com/masters3d> I've rebased and tested the
> changes locally, but I can't push to your branch. Please try adding me as a
> collaborator again.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apple/swift/pull/16577#discussion_r205973877>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AGPIlD3u0u83YTmyMjWggghY4n9bMb81ks5uLbkKgaJpZM4T8mdN>
> .
>

masters3d commented:
Thanks @benrimmington I reset my branch to yours.

benrimmington commented:
@airspeedswift Please could you trigger swift-ci testing again.

natecook1000 commented:
@swift-ci Please smoke test

benrimmington commented:
@airspeedswift The smoke test has finished. Can this be merged? There have been a few extra commits, since you last approved this.

benrimmington commented:
@masters3d Please reset your branch to [`benrimmington:renamed-to-KeyValueArray`](https://github.com/benrimmington/swift/tree/renamed-to-KeyValueArray), to resolve another conflict.

airspeedswift commented:
@swift-ci please test

airspeedswift commented:
@swift-ci please clean test

benrimmington commented:
The failure is due to a missing type change in [test/api-digester/source-stability.swift.expected](https://github.com/apple/swift/blob/master/test/api-digester/source-stability.swift.expected)

```diff
 Var Dictionary.values has declared type change from LazyMapCollection<[Key : Value], Value> to Dictionary<Key, Value>.Values
+Constructor Mirror.init(_:children:displayStyle:ancestorRepresentation:) has parameter 1 type change from DictionaryLiteral<String, Any> to KeyValuePairs<String, Any>
 Constructor String.init(_:) has return type change from String? to String
```

airspeedswift commented:
@akyrtzi should that api difference be firing? the one is a typealias of the other.

airspeedswift commented:
I mean, it'd be an ABI break once we stabilize the ABI, but it shouldn't be a source compatibility break.

nkcsgexi commented:
Can we update the expected change list for now to unblock the merging of this patch?

airspeedswift commented:
sure, just want to check if it's a bug that it flagged.

nkcsgexi commented:
yeah, i think the api checker currently isn't smart enough to tell a type alias change is benign for source compatibility. This won't be an issue when we check abi stability because abi checker compares module dumps with all types canonicalized.

airspeedswift commented:
Hi @masters3d can you add that entry to the expected results to fix the test for now? Thanks!

benrimmington committed:
[SE-0214] Update source-stability.swift.expected

benrimmington commented:
@masters3d I've updated [`benrimmington:renamed-to-KeyValueArray`](https://github.com/benrimmington/swift/tree/renamed-to-KeyValueArray) again, if you'd prefer to reset your branch.

masters3d committed:
Merge branch 'master' into renamed-to-KeyValueArray

masters3d commented:
@airspeedswift @benrimmington updated

airspeedswift commented:
@swift-ci please test

airspeedswift commented:
@swift-ci please test source compatibility

benrimmington commented:
@airspeedswift You'll probably need a "clean test", because my branch had been rebased.

@masters3d I've rebased again for a merge conflict, but it looks like you've already fixed it?

airspeedswift commented:
Those failures are spurious CI artifacts, not actual failures.

airspeedswift commented:
@swift-ci please test macOS platform

airspeedswift commented:
@shahmishal any idea why macOS tests aren't triggering on this PR?

shahmishal commented:
It's running - https://ci.swift.org/job/swift-PR-osx/7078/

benrimmington commented:
Lines 478...480 of Mirror.swift could be re-indented (in a follow-up PR).

benrimmington/swift@8aa413ccbceef56faeaf3d60ef5a3132af1e2581

masters3d commented:
Thank you swift team! Thanks @benrimmington !
